tools/subtype-gen/rules_parser.go

```diff
@@ -0,0 +1,459 @@
+		panic(fmt.Errorf(
+			"type name %#[1]q is not suffixed with %#[2]q."+
+				" Replace the type-name with `%[1]s%[2]s`",
+			typePlaceHolder,
+			TypePlaceholderSuffix,
+		))
+	}
+
+	typeName := strings.TrimSuffix(typePlaceHolder, TypePlaceholderSuffix)
+
+	switch typeName {
+	case TypePlaceholderOptional,
+		TypePlaceholderDictionary,
+		TypePlaceholderVariableSized,
+		TypePlaceholderConstantSized,
+		TypePlaceholderReference,
+		TypePlaceholderComposite,
+		TypePlaceholderInterface,
+		TypePlaceholderFunction,
+		TypePlaceholderIntersection,
+		TypePlaceholderParameterized,
+		TypePlaceholderConforming:
+		return ComplexType{
+			name: typeName,
+		}
+	default:
+		return SimpleType{
+			name: typeName,
+		}
+	}
+}
+
+func parseMemberExpression(names []string) Expression {
+	size := len(names)
+	if size != 2 {
+		panic(fmt.Errorf("invalid number of nested levels for member: expected 2, found %d", size))
+	}
+
+	return MemberExpression{
+		Parent: IdentifierExpression{
+			Name: names[0],
+		},
+		MemberName: names[1],
+	}
+}
+
+// parsePredicate parses a predicate from the YAML
+func parsePredicate(predicate any) (Predicate, error) {
+	switch v := predicate.(type) {
+	case string:
+		switch v {
+		case "always":
+			return AlwaysPredicate{}, nil
+		case "never":
+			return NeverPredicate{}, nil
+		default:
+			return nil, fmt.Errorf("unsupported string rule: %s", v)
+		}
+	case KeyValues:
+		key, value := singleKeyValueFromMap(v)
+
+		switch key {
+
+		case "isResource":
+			expr := parseSimpleExpression(value)
+			return IsResourcePredicate{Expression: expr}, nil
+
+		case "isAttachment":
+			expr := parseSimpleExpression(value)
+			return IsAttachmentPredicate{Expression: expr}, nil
+
+		case "isHashableStruct":
+			expr := parseSimpleExpression(value)
+			return IsHashableStructPredicate{Expression: expr}, nil
+
+		case "isStorable":
+			expr := parseSimpleExpression(value)
+			return IsStorablePredicate{Expression: expr}, nil
+
+		case "equals":
+			sourceExpr, targetExpr, err := parseSourceAndTarget(key, value)
+			if err != nil {
+				return nil, err
+			}
+
+			return EqualsPredicate{
+				Source: sourceExpr,
+				Target: targetExpr,
+			}, nil
+
+		case "subtype":
+			superType, subType, err := parseSuperAndSubExpressions(key, value)
+			if err != nil {
+				return nil, err
+			}
+
+			return SubtypePredicate{
+				Sub:   subType,
+				Super: superType,
+			}, nil
+
+		case "and":
+			and, ok := value.([]any)
+			if !ok {
+				return nil, fmt.Errorf("expected []any, got %T", value)
+			}
+
+			var predicates []Predicate
+			for _, cond := range and {
+				predicate, err := parsePredicate(cond)
+				if err != nil {
+					return nil, err
+				}
+				predicates = append(predicates, predicate)
+			}
+
+			return AndPredicate{Predicates: predicates}, nil
+
+		case "or":
+			or, ok := value.([]any)
+			if !ok {
+				return nil, fmt.Errorf("expected a list of predicates, got %T", value)
```

Comment:
Future idea: Improved error reporting, e.g. by keeping a path when parsing predicates (and rules), similar to #4302. Especially when rules get complicated it might be hard to find in the point in the rules files this error is for.

Reply:
Good idea! Added an item here: https://github.com/onflow/cadence/issues/3691#issuecomment-3439450533